core/tests/ledger_cleanup.rs

```diff
@@ -0,0 +1,296 @@
+            self.cpu_stats.get_stats()
+        }
+
+        pub fn join(self) -> std::thread::Result<()> {
+            self.t_cleanup.join()
+        }
+    }
+
+    fn read_env<T>(key: &str, default: T) -> T
+    where
+        T: FromStr,
+    {
+        match std::env::var(key) {
+            Ok(val) => val.parse().unwrap_or(default),
+            Err(_e) => default,
+        }
+    }
+
+    fn get_benchmark_config() -> BenchmarkConfig {
+        let benchmark_slots = read_env("BENCHMARK_SLOTS", DEFAULT_BENCHMARK_SLOTS);
+        let batch_size = read_env("BATCH_SIZE", DEFAULT_BATCH_SIZE);
+        let max_ledger_slots = read_env("MAX_LEDGER_SLOTS", DEFAULT_MAX_LEDGER_SLOTS);
+        let cleanup_blocktree = read_env("CLEANUP_BLOCKTREE", true);
+        let emit_cpu_info = read_env("EMIT_CPU_INFO", true);
+        // set default to `true` once compaction is merged
+        let assert_compaction = read_env("ASSERT_COMPACTION", false);
+
+        BenchmarkConfig {
+            benchmark_slots,
+            batch_size,
+            max_ledger_slots,
+            cleanup_blocktree,
+            emit_cpu_info,
+            assert_compaction,
+        }
+    }
+
+    fn emit_header() {
+        println!("EPOCH_MS,DELTA_MS,START_SLOT,BATCH_SIZE,ENTRIES,MAX,SIZE,DELTA_SIZE,CPU_USER,CPU_SYSTEM,CPU_IDLE");
+    }
+
+    fn emit_stats(
+        time_previous: &mut SystemTime,
+        storage_previous: &mut u64,
+        start_slot: u64,
+        batch_size: u64,
+        entries: u64,
+        max_slots: i64,
+        blocktree: &Blocktree,
+        cpu: &CpuStatsInner,
+    ) {
+        let time_now = SystemTime::now();
+        let storage_now = blocktree.storage_size().unwrap_or(0);
+        let (cpu_user, cpu_system, cpu_idle) = (cpu.cpu_user, cpu.cpu_system, cpu.cpu_idle);
+
+        println!(
+            "{},{},{},{},{},{},{},{},{},{},{}",
+            time_now
+                .duration_since(SystemTime::UNIX_EPOCH)
+                .unwrap()
+                .as_millis(),
+            time_now.duration_since(*time_previous).unwrap().as_millis(),
+            start_slot,
+            batch_size,
+            entries,
+            max_slots,
+            storage_now,
+            storage_now as i64 - *storage_previous as i64,
+            cpu_user,
+            cpu_system,
+            cpu_idle,
+        );
+
+        *time_previous = time_now;
+        *storage_previous = storage_now;
+    }
+
+    #[test]
+    fn test_ledger_cleanup_compaction() {
+        let blocktree_path = get_tmp_ledger_path!();
+        let blocktree = Arc::new(Blocktree::open(&blocktree_path).unwrap());
+        let config = get_benchmark_config();
+        eprintln!("BENCHMARK CONFIG: {:?}", config);
+        eprintln!("LEDGER_PATH: {:?}", &blocktree_path);
+
+        let benchmark_slots = config.benchmark_slots;
+        let batch_size = config.batch_size;
+        let max_ledger_slots = config.max_ledger_slots;
+        let batches = benchmark_slots / batch_size;
```

Comment:
Can we parameterize this by size in bytes? That way it makes it easier to configure scenarios like:

1) Writing enough data to fill up a write buffer
2) Writing enough data to see SST's generated from L0 -> L1
3) Writing enough data to see SST's generated from L1->L2-> ... -> L_n

Reply:
I like this idea! To avoid arithmetic over quantities I don't know or which
are subject to change, I'm thinking of implementing this as
a configurable `STOP_SIZE_BYTES` and `STOP_SIZE_ITERATIONS` parameters,
such that the benchmark will terminate when rocksdb directory size is
`STOP_SIZE_BYTES` for `STOP_SIZE_ITERATIONS`.

(i.e. exceed 1GB for 5 iterations)

Reply:
(added in latest version)